### PR TITLE
Fluidstack integration

### PR DESCRIPTION
**Pull request by mjibril**
- Added FluidStack Cloud Integration
 - List accelerators and prices
 - Launch clusters on FluidStack

Tested (run the relevant ones):

- [ x] Code formatting: `bash format.sh`
- [x ] Any manual or new tests for this PR (please specify below)
   Fetched and saved FluidStack catalog. Launched and terminated clusters on FluidStack. 

- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Comment by Michaelvll** [comment]
Thanks for submitting the PR @mjibril! It is very exciting to see the fluidstack support in SkyPilot. We will go through the PR soon. Just a note, we prefer to use single quote, instead of double across the project.
Could you try to `pip install -r requirements-dev.txt` and run `./format.sh`?

**Comment by mjibril** [comment]
@Michaelvll replaced double with single quotes.

**Comment by Michaelvll** [comment]
> @Michaelvll replaced double with single quotes.

Thanks @mjibril! Could you check if the changes are pushed? It seems the 6267af9 commit above does not include the changes. : )

File: sky/authentication.py [hunk 1]
```diff
@@ -42,6 +42,7 @@
 from sky.utils import subprocess_utils
```
Comment (Michaelvll) [inline sky/authentication.py:42]: Did we accidently format this file with a different format setting?

File: sky/backends/cloud_vm_ray_backend.py [hunk 2]
```diff
@@ -87,61 +87,69 @@
 _FETCH_IP_MAX_ATTEMPTS = 3
```
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:87]: nit: same quote issue?

File: sky/clouds/fluidstack.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    from sky import resources as resources_lib
+
+
+@clouds.CLOUD_REGISTRY.register
+class Fluidstack(clouds.Cloud):
+    """FluidStack GPU Cloud."""
+
+    _REPR = 'Fluidstack'
+
+    # Lamdba has a 64 char limit for cluster name.
```
Comment (Michaelvll) [inline sky/clouds/fluidstack.py:35]:
```suggestion
    # Fluidstack has a 64 char limit for cluster name.
```

File: sky/clouds/fluidstack.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    # Reference: https://cloud.lambdalabs.com/api/v1/docs#operation/launchInstance # pylint: disable=line-too-long
```
Comment (Michaelvll) [inline sky/clouds/fluidstack.py:36]: Need to update the comment?

File: sky/clouds/fluidstack.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+        **kwargs,
+    ) -> List[status_lib.ClusterStatus]:
+        status_map = {
+            "Provisioning": status_lib.ClusterStatus.INIT,
+            "Starting Up": status_lib.ClusterStatus.INIT,
+            "Running": status_lib.ClusterStatus.UP,
+            "Error Creating": status_lib.ClusterStatus.INIT,  #should be error
+        }
+        status_list = []
+        vms = FluidstackClient().list_instances()
```
Comment (Michaelvll) [inline sky/clouds/fluidstack.py:296]: Do we need to filter out the vms with the `name` or the `tag_filters`?

File: sky/clouds/service_catalog/data_fetchers/fetch_fluidstack_cloud.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
+"""A script that generates the Lambda Cloud catalog.
```
Comment (Michaelvll) [inline sky/clouds/service_catalog/data_fetchers/fetch_fluidstack_cloud.py:1]: This is fantastic!

File: sky/clouds/fluidstack.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+    FLUIDSTACK_API_KEY_PATH,
+    FLUIDSTACK_API_TOKEN_PATH,
+    FluidstackClient,
+)
+import os
+
+_CREDENTIAL_FILES = [
+    # credential files for FluidStack,
+    FLUIDSTACK_API_KEY_PATH,
+    FLUIDSTACK_API_TOKEN_PATH,
```
Comment (Michaelvll) [inline sky/clouds/fluidstack.py:22]: nit: import modules instead of variables.

**Review by Michaelvll — COMMENTED** [review]
Thanks for submitting the PR @mjibril! This is fantastic! I just went through the PR and really enjoy the simplicity of the Fluidstack API! Left several initial comments, and will test it out soon!

It seems the quote changes are still there, which makes it a bit hard to see the actual code changes. Could you check if your fixes were correctly committed and pushed?

File: sky/clouds/fluidstack.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+    
+        try:
+            assert os.path.exists(os.path.expanduser(FLUIDSTACK_API_KEY_PATH))
+            assert os.path.exists(os.path.expanduser(FLUIDSTACK_API_TOKEN_PATH))
+        except AssertionError:
+            return False, (
+                'Failed to access FluidStack Cloud with credentials. '
+                'To configure credentials, go to:\n    '
+                '  https://console.fluidstack.io \n    '
+                'to obtain an API key and API Token, then add save the contents\n'
```
Comment (Michaelvll) [inline sky/clouds/fluidstack.py:238]: We may need indents here as well?

**Comment by Michaelvll** [comment]
For testing, could you help merge the latest changes from the master branch @mjibril : )

**Review by Michaelvll — COMMENTED** [review]
I am trying to run the following command but get an error:
```
sky launch --cloud fluidstack --gpus A10 -c test-fs
```

```
Traceback (most recent call last):
2023-09-04 23:38:18,721 INFO commands.py:385 -- Checking External environment settings
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/requests/models.py", line 971, in json
2023-09-04 23:38:21,820 INFO commands.py:686 -- No head node found. Launching a new cluster. Confirm [y/N]: y [automatic, due to --yes]
2023-09-04 23:38:21,820 INFO usage_lib.py:381 -- Usage stats collection is disabled.
No head node exists, need to create it.
2023-09-04 23:38:21,820 INFO commands.py:743 -- Acquiring an up-to-date head node
    return complexjson.loads(self.text, **kwargs)
  File "/opt/conda/envs/sky/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/opt/conda/envs/sky/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/opt/conda/envs/sky/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/gcpuser/skypilot/sky/skylet/providers/fluidstack/fluidstack_utils.py", line 36, in raise_fluidstack_error
    resp_json = response.json()
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/requests/models.py", line 975, in json
    raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/tmp/skypilot_ray_up_g_iuhk31.py", line 76, in <module>
    sdk.create_or_update_cluster('/home/gcpuser/.sky/generated/test-fs.yml', **{'no_restart': True})
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/ray/autoscaler/sdk/sdk.py", line 38, in create_or_update_cluster
    return commands.create_or_update_cluster(
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/ray/autoscaler/_private/commands.py", line 314, in create_or_update_cluster
    get_or_create_head_node(
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/ray/autoscaler/_private/commands.py", line 758, in get_or_create_head_node
    provider.create_node(head_node_config, head_node_tags, 1)
  File "/home/gcpuser/skypilot/sky/skylet/providers/fluidstack/node_provider.py", line 133, in create_node
    instance = self.fluidstack_client.info(vm_id)
  File "/home/gcpuser/skypilot/sky/skylet/providers/fluidstack/fluidstack_utils.py", line 160, in info
    raise_fluidstack_error(response)
  File "/home/gcpuser/skypilot/sky/skylet/providers/fluidstack/fluidstack_utils.py", line 39, in raise_fluidstack_error
    raise FluidstackAPIError(
sky.skylet.providers.fluidstack.fluidstack_utils.FluidstackAPIError: Unexpected error. Status code: 503
```

When we try to terminate the cluster, the following failure occurs:
```console
> sky down test-fs
Terminating 1 cluster: test-fs. Proceed? [Y/n]:
Terminating cluster test-fs...failed.
Reason: builtins.RuntimeError:
Failed to terminate test-fs. If you want to ignore this error and remove the cluster from the status table, use `sky down --purge`.
**** STDOUT ****
2023-09-04 23:45:36,153 VWARN commands.py:354 -- Loaded cached provider configuration from /var/tmp/ray-config-28d6143eb919c25343c2274473838bcbd0f99159
2023-09-04 23:45:36,153 WARN commands.py:360 -- If you experience issues with the cloud provider, try re-running the command with --no-config-cache.
2023-09-04 23:45:36,153 INFO commands.py:440 -- Destroying cluster. Confirm [y/N]: y [automatic, due to --yes]
2023-09-04 23:45:38,215 VERR commands.py:458 -- 'str' object has no attribute 'get'
2023-09-04 23:45:38,215 WARN commands.py:459 -- Exception occurred when stopping the cluster Ray runtime (use -v to dump teardown exceptions).
2023-09-04 23:45:38,215 WARN commands.py:463 -- Ignoring the exception and attempting to shut down the cluster nodes anyway.

**** STDERR ****
2023-09-04 23:45:36,152 WARNING util.py:252 -- Dropping the empty legacy field head_node. head_nodeis not supported for ray>=2.0.0. It is recommended to removehead_node from the cluster config.
2023-09-04 23:45:36,152 WARNING util.py:252 -- Dropping the empty legacy field worker_nodes. worker_nodesis not supported for ray>=2.0.0. It is recommended to removeworker_nodes from the cluster config.
2023-09-04 23:45:36,152 INFO util.py:375 -- setting max workers for head node type to 0
2023-09-04 23:45:36,168 WARNING util.py:252 -- Dropping the empty legacy field head_node. head_nodeis not supported for ray>=2.0.0. It is recommended to removehead_node from the cluster config.
2023-09-04 23:45:36,168 WARNING util.py:252 -- Dropping the empty legacy field worker_nodes. worker_nodesis not supported for ray>=2.0.0. It is recommended to removeworker_nodes from the cluster config.
2023-09-04 23:45:36,168 INFO util.py:375 -- setting max workers for head node type to 0
Traceback (most recent call last):
  File "/opt/conda/envs/sky/bin/ray", line 8, in <module>
    sys.exit(main())
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/ray/scripts/scripts.py", line 2474, in main
    return cli()
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/ray/autoscaler/_private/cli_logger.py", line 856, in wrapper
    return f(*args, **kwargs)
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/ray/scripts/scripts.py", line 1313, in down
    teardown_cluster(
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/ray/autoscaler/_private/commands.py", line 526, in teardown_cluster
    A = remaining_nodes()
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/ray/autoscaler/_private/commands.py", line 471, in remaining_nodes
    workers = provider.non_terminated_nodes({TAG_RAY_NODE_KIND: NODE_KIND_WORKER})
  File "/home/gcpuser/skypilot/sky/skylet/providers/fluidstack/node_provider.py", line 73, in non_terminated_nodes
    nodes = self._get_filtered_nodes(tag_filters=tag_filters)
  File "/home/gcpuser/skypilot/sky/skylet/providers/fluidstack/node_provider.py", line 60, in _get_filtered_nodes
    if instance["tags"].get(key,None) == value:
AttributeError: 'str' object has no attribute 'get'
.
```

**Comment by mjibril** [comment]
@Michaelvll Can you check the PR again with the latest push?

**Comment by Michaelvll** [comment]
Thanks for the quick fix @mjibril! I am trying out some basic tests:
1. `sky launch --cloud fluidstack --cpus 2+`, it seems stuck in the launching phase (in `provision.log` there is only `setting max workers for head node type to 0`). I checked the console, and it seems the server was "starting up" for a while and disappeared. Note that `RTX3090` was used.
2. `sky launch --cloud fluidstack --gpus A10`. It seems I get the `sky.skylet.providers.fluidstack.fluidstack_utils.FluidstackAPIError: Unexpected error. Status code: 503` again in the `provision.log`. However, on the console, the machine is UP.
3. `sky launch --cloud fluidstack --gpus A100-80GB`. The issue as 1 happens.

File: sky/skylet/providers/fluidstack/node_provider.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        if not tag_filters:
+            for instance in running_instances:
+                self.cached_nodes[instance['id']] = instance
+            return self.cached_nodes
+
+        for instance in running_instances:
+            for key, value in tag_filters.items():
+                tag = instance['tags'].get(key, None)
+                if tag and tag == value:
+                    self.cached_nodes[instance['id']] = instance
```
Comment (Michaelvll) [inline sky/skylet/providers/fluidstack/node_provider.py:55]: Seems we have this tag filtering code in many places. Do we want to add it to the API of `fluidstack_client.list_instances`, i.e. `list_instances(tag_filters: Optional[Dict[str, str]])`

File: sky/clouds/service_catalog/data_fetchers/fetch_fluidstack_cloud.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    'Quadro_RTX_5000_16GB': 'RTX5000',
+    'A30': 'A30',
+    'A40': 'A40',
+    'RTX_3090': 'RTX3090',
+    'RTX_A6000': 'A6000',
+    'RTX_3080': 'RTX3080',
+    'RTX_A5000': 'A5000',
+    'A100_SXM4_80GB': 'A100-80GB',
+    'A100_NVLINK_80GB': 'A100-80GB-NVLink',
+    'A100_NVLINK': 'A100_NVLINK'
```
Comment (Michaelvll) [inline sky/clouds/service_catalog/data_fetchers/fetch_fluidstack_cloud.py:45]: Let's avoid adding the `NVLink` to the accelerator name, but just keep them `A100` and `A100-80GB`. Our higher level code will handle the case when multiple instance types are involved for the same accelerator, hinting the user to explicitly specify the instance type.

File: sky/utils/accelerator_registry.py [hunk 11]
```diff
@@ -37,6 +37,21 @@
...
+    'A4000',
+    'RTX5000',
+    'A30',
+    'A40',
+    'RTX3090',
+    'A6000',
+    'RTX3080',
+    'A5000',
+    'A100-80GB-NVLink',
+    'A100_NVLINK',
```
Comment (Michaelvll) [inline sky/utils/accelerator_registry.py:54]: Would be nice to remove : )

File: sky/clouds/fluidstack.py [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+        clouds.CloudImplementationFeatures.SPOT_INSTANCE:
+            'Spot instances are'
+            f' not supported in {_REPR}.',
+        clouds.CloudImplementationFeatures.CUSTOM_DISK_TIER:
+            'Custom disk tiers'
+            f' are not supported in {_REPR}.',
+        clouds.CloudImplementationFeatures.OPEN_PORTS:
+            'Opening ports'
+            f'is not supported in {_REPR}.',
+    }
```
Comment (Michaelvll) [inline sky/clouds/fluidstack.py:58]: Should we add `clouds.CloudImplementationFeatures.MULTI_NODE`? It seems we disable it in the `node_provider`?

**Review by Michaelvll — COMMENTED** [review]
Thank you for submitting the PR for supporting Fluidstack @mjibril! Very excited to see the support. It works for me to launch a cluster with RTX3090 and I will test more with this PR.

Several TODOs:
- [ ] Add some tests in https://github.com/skypilot-org/skypilot/blob/f2e6f1e99560d287bd2f47c7ba912b7333c759bd/tests/test_smoke.py#L1. Please refer to the changes for `oci` in https://github.com/skypilot-org/skypilot/blob/f2e6f1e99560d287bd2f47c7ba912b7333c759bd/tests/conftest.py#L25

File: sky/clouds/service_catalog/data_fetchers/fetch_fluidstack_cloud.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
+"""A script that generates the Lambda Cloud catalog.
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/clouds/service_catalog/data_fetchers/fetch_fluidstack_cloud.py:1]: Also, I would prefer naming the file as `fetch_fluidstack`. We named Lambda as `lambda_cloud` to distinguish it with the `AWS Lambda` : )

**Review by Michaelvll — COMMENTED** [review]
For termination and stopping of the cluster, please checkout our latest implementation for AWS and GCP: 
https://github.com/skypilot-org/skypilot/blob/f2e6f1e99560d287bd2f47c7ba912b7333c759bd/sky/provision/aws/instance.py#L132-L175
https://github.com/skypilot-org/skypilot/blob/f2e6f1e99560d287bd2f47c7ba912b7333c759bd/sky/provision/gcp/instance.py#L132-L175

We are deprecating the usage of ray for the termination of clusters, as it can cause some leakage of the VM if they are in some certain states. Instead, we are now directly calling the cloud API to terminate the VM. After that is implemented, please remember to update the list of cloud to use the new termination here: https://github.com/skypilot-org/skypilot/blob/f2e6f1e99560d287bd2f47c7ba912b7333c759bd/sky/backends/cloud_vm_ray_backend.py#L3720

**Comment by Michaelvll** [comment]
Fluidstack has been supported by #3086. Closing this PR now. : )

**Pull request closed without merging**